### PR TITLE
Add option for codecov and allocation tracking to be restricted by path

### LINKED ISSUE #36142: Coverage test is extremely slow on multiple threads
Running `Pkg.test("MyModule", coverage=true)` on more than one thread is extremely slow.

I measured the time on my machine using 1, 2, 8 and 16 threads:

1 thread: 339 seconds
2 threads: 3106 seconds
8 threads: 6401 seconds
16 threads: 5072 seconds

Running `Pkg.test("MyModule")` without coverage takes approximately the same time for every number of threads (around 2 minutes) for this module.

See also the [discourse discussion](https://discourse.julialang.org/t/coverage-test-is-extremely-slow-on-multiple-threads/40682).

I am using Julia 1.4.1 on Windows 10 and my CPU has 8 cores/16 threads.

### LINKED ISSUE #41626: Limit code coverage assessment to within a given directory for efficiency
Code coverage has `--code-coverage=user` but for CI most of the time you only care about coverage within the repo being tested.

Code coverage also slows execution down a fair bit.

Would it be possible to add the option to `--code-coverage=/path/to/repo` in such a way that would make it more efficient?

One thought was if the source file info was within the `linetable`? That could make it relatively simple to extend functionality to `do_coverage`, perhaps?
https://github.com/JuliaLang/julia/blob/56bb340414f7d7015576b74ac7edbc24ffbaa15f/src/codegen.cpp#L6935

### PR DESCRIPTION
**Pull request by IanButterworth**
Replaces https://github.com/JuliaLang/julia/pull/43711 (triage suggested this approach as cleaner, which I agree with)
Closes https://github.com/JuliaLang/julia/issues/41626
Closes https://github.com/JuliaLang/julia/issues/36142

Enabling code coverage and allocation tracking often blows up test suite duration.
The narrowest option currently is `user` which is all non-core code, including stdlibs, and given that code coverage tools usually only look for coverage in the checked out directory, the vast majority of effort is usually needless.

This PR introduces a `--code-coverage=@/path/to/package` and the same for `--track-allocation` option to declare tracking of anything falling in or under the given path. i.e. can be a specific file or a directory.

Testing use of LoopVectorization, which seems particularly effected.
Quoting @chriselrod [here](https://github.com/JuliaPerf/STREAMBenchmark.jl/pull/15#issuecomment-958749496)
> coverage=true also seems to cause single-threaded `@turbo` to be very slow.
> Yet coverage=false tests take about 10-12 minutes on CI, and many of the coverage=true tests take 2+ hours

Taking a simple LoopVectorization example:
```julia
function mydotavx(a, b)
    s = 0.0
    @turbo for i ∈ eachindex(a,b)
        s += a[i]*b[i]
    end
    s
end
```

```
% julia --code-coverage=none --project  -e "using Foo, BenchmarkTools; a = rand(256); b = rand(256); @btime Foo.mydotavx(\$a, \$b)"
  22.645 ns (0 allocations: 0 bytes)

% julia --code-coverage=@/Users/ian/Documents/GitHub/Foo.jl --project  -e "using Foo, BenchmarkTools; a = rand(256); b = rand(256); @btime Foo.mydotavx(\$a, \$b)"
  26.826 ns (0 allocations: 0 bytes)

% julia --code-coverage=user --project  -e "using Foo, BenchmarkTools; a = rand(256); b = rand(256); @btime Foo.mydotavx(\$a, \$b)"
  532.074 ns (0 allocations: 0 bytes)

% julia --code-coverage=all --project  -e "using Foo, BenchmarkTools; a = rand(256); b = rand(256); @btime Foo.mydotavx(\$a, \$b)"
  633.341 ns (0 allocations: 0 bytes)
```

The effects are stronger for `--track-allocation` for code that heavily allocates. 

I'm disappointed I didn't get this done before the 1.8 freeze, as this issue prevents us running code coverage on some long-running production test suites.

cc. @efaulhaber @carstenbauer

### CODE REVIEW

**Comment by IanButterworth** [comment]
The new coverage tests are failing due to a slight difference (the expected report is on top)
```
"SF:/Users/ian/Documents/GitHub/julia/test/testhelpers/coverage_file.jl\nDA:3,1\nDA:4,1\nDA:5,0\nDA:7,1\nDA:8,1\nDA:9,3\nDA:10,5\nDA:11,1\nDA:12,1\nDA:14,0\nDA:17,1\nDA:18,0\nDA:19,2\nDA:20,1\nDA:22,1\nDA:1234,0\nLH:12\nLF:16\nend_of_record\n", 
"SF:/Users/ian/Documents/GitHub/julia/test/testhelpers/coverage_file.jl\nDA:3,1\nDA:4,1\nDA:5,0\nDA:7,1\nDA:8,1\nDA:9,3\nDA:10,5\nDA:11,0\nDA:12,1\nDA:14,0\nDA:17,1\nDA:18,0\nDA:19,0\nDA:20,0\nDA:22,1\nDA:1234,0\nLH:9\nLF:16\nend_of_record\n"
```
I'm not sure what would cause a subtle difference like this

**Comment by IanButterworth** [comment]
Perhaps if approved, this could go in to 1.8 as strictly opt-in. i.e. `Pkg.test(.. coverage=true)` still would default to `user` in 1.8 but this could still be enabled via setting an overriding julia arg. 

If it works well in the wild, it could become default in 1.9?

File: src/codegen.cpp [hunk 1]
```diff
@@ -6977,15 +6977,20 @@ static std::pair<std::unique_ptr<Module>, jl_llvm_functions_t>
         return (!jl_is_submodule(mod, jl_base_module) &&
                 !jl_is_submodule(mod, jl_core_module));
     };
+    auto in_tracked_path = [] (StringRef file) {
+        return (jl_options.tracked_path != NULL && file.rfind(jl_options.tracked_path) == 0);
```
Comment (vtjnash) [inline src/codegen.cpp:6981]:
```suggestion
        return jl_options.tracked_path != NULL && file.startsWith(jl_options.tracked_path);
```

File: src/codegen.cpp [hunk 2]
```diff
@@ -7011,13 +7017,16 @@ static std::pair<std::unique_ptr<Module>, jl_llvm_functions_t>
...
             info.file = jl_symbol_name(filesym);
             if (info.file.empty())
                 info.file = "<missing>";
+            if (module == ctx.module) {
+                info.is_user_code = mod_is_user_mod;
+                info.is_tracked = mod_is_tracked;
+            } else {
+                info.is_user_code = in_user_mod(module);
+                info.is_tracked = in_tracked_path(info.file);
+            }
```
Comment (vtjnash) [inline src/codegen.cpp:7029]: the module and file aren't necessarily related
```suggestion
            if (module == ctx.module)
                info.is_user_code = mod_is_user_mod;
            else
                info.is_user_code = in_user_mod(module);
            info.is_tracked = in_tracked_path(info.file);
```

File: src/jloptions.c [hunk 3]
```diff
@@ -520,6 +525,10 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
                         codecov = JL_LOG_ALL;
                     jl_options.output_code_coverage = optarg;
                 }
+                else if (!strncmp(optarg, "@", 1)) {
+                    codecov = JL_LOG_PATH;
+                    jl_options.tracked_path = memmove(optarg, optarg+1, strlen(optarg)); // remove `@`
```
Comment (vtjnash) [inline src/jloptions.c:530]:
```suggestion
                    jl_options.tracked_path = optarg + 1; // skip `@`
```

File: src/jloptions.c [hunk 4]
```diff
@@ -536,6 +545,10 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
                     malloclog = JL_LOG_ALL;
                 else if (!strcmp(optarg,"none"))
                     malloclog = JL_LOG_NONE;
+                else if (!strncmp(optarg, "@", 1)) {
+                    malloclog = JL_LOG_PATH;
+                    jl_options.tracked_path = memmove(optarg, optarg+1, strlen(optarg)); // remove `@`
```
Comment (vtjnash) [inline src/jloptions.c:550]:
```suggestion
                    jl_options.tracked_path = optarg + 1; // remove `@`
```

File: src/jloptions.c [hunk 5]
```diff
@@ -154,11 +155,15 @@ static const char opts[]  =
...
     " --code-coverage[={none*|user|all}]\n"
     "                            Count executions of source lines (omitting setting is equivalent to `user`)\n"
+    " --code-coverage@/abspath\n"
+    "                            Count executions but only in files that fall under the given absolute file path/directory\n"
     " --code-coverage=tracefile.info\n"
     "                            Append coverage information to the LCOV tracefile (filename supports format tokens)\n"
 // TODO: These TOKENS are defined in `runtime_ccall.cpp`. A more verbose `--help` should include that list here.
     " --track-allocation[={none*|user|all}]\n"
     "                            Count bytes allocated by each source line (omitting setting is equivalent to `user`)\n"
+    " --track-allocation@/abspath\n"
```
Comment (vtjnash) [inline src/jloptions.c:165]:
```suggestion
    " --track-allocation=@/abspath\n"
```

File: src/jloptions.c [hunk 6]
```diff
@@ -154,11 +155,15 @@ static const char opts[]  =
     // instrumentation options
     " --code-coverage[={none*|user|all}]\n"
     "                            Count executions of source lines (omitting setting is equivalent to `user`)\n"
+    " --code-coverage@/abspath\n"
```
Comment (vtjnash) [inline src/jloptions.c:158]:
```suggestion
    " --code-coverage=@/abspath\n"
```

File: base/util.jl [hunk 7]
```diff
@@ -196,6 +196,8 @@ function julia_cmd(julia=joinpath(Sys.BINDIR, julia_exename()))
                 push!(addflags, "--code-coverage=user")
             elseif opts.code_coverage == 2
                 push!(addflags, "--code-coverage=all")
+            elseif opts.code_coverage == 3
+                push!(addflags, "--code-coverage=@$(opts.tracked_path)")
```
Comment (vtjnash) [inline base/util.jl:200]:
```suggestion
                push!(addflags, "--code-coverage=@$(unsafe_string(opts.tracked_path))")
```

File: src/codegen.cpp [hunk 8]
```diff
@@ -6977,15 +6977,20 @@ static std::pair<std::unique_ptr<Module>, jl_llvm_functions_t>
         return (!jl_is_submodule(mod, jl_base_module) &&
                 !jl_is_submodule(mod, jl_core_module));
     };
+    auto in_tracked_path = [] (StringRef file) {
+        return jl_options.tracked_path != NULL && file.startsWith(jl_options.tracked_path);
```
Comment (IanButterworth) [inline src/codegen.cpp:6981]:
```suggestion
        return jl_options.tracked_path != NULL && file.startswith(jl_options.tracked_path);
```

File: src/jloptions.c (hunk 4 above)
Comment (IanButterworth) [reply src/jloptions.c:550]:
```suggestion
                    jl_options.tracked_path = optarg + 1; // skip `@`
```

File: src/jloptions.c [hunk 9]
```diff
@@ -154,11 +155,18 @@ static const char opts[]  =
     // instrumentation options
     " --code-coverage[={none*|user|all}]\n"
     "                            Count executions of source lines (omitting setting is equivalent to `user`)\n"
+    " --code-coverage=@/abs/path\n"
```
Comment (fredrikekre) [inline src/jloptions.c:158]: I don't think any other file input requires an absolute path, couldn't the path be absolutified instead?

Comment (IanButterworth) [reply src/jloptions.c]: Good point! I was unaware how the paths were absolutified. 

The nice thing now is that a simple `--code-coverage=@` will track the current directory

**Pull request merged**